### PR TITLE
Regex validation for date of birth claim value

### PR DESCRIPTION
**Pull request by AnuradhaSK**
## Purpose
When the local claim of dob ("http://wso2.org/claims/dob") is considered as a string in the SCIM schema after https://github.com/wso2-extensions/identity-inbound-provisioning-scim2/pull/330 is merged. 
and currently, there is no regex configured for local dob claim.

Here we validate the dob against the configured regex or the default regex : `^\d{4}-\d{2}-\d{2}$`, in SCIMUserOperationListener 
when adding a user and updating a user

Resolves part of https://github.com/wso2/product-is/issues/11226

### CODE REVIEW

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 1]
```diff
@@ -165,6 +169,12 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
         Map<String, String> scimToLocalMappings = SCIMCommonUtils.getSCIMtoLocalMappings();
         String modifiedLocalClaimUri = scimToLocalMappings.get(SCIMConstants.CommonSchemaConstants.LAST_MODIFIED_URI);
         claims.put(modifiedLocalClaimUri, lastModifiedDate);
+
+        // Change dateOfBirth value to datetime format if dob is in yyyy-mm-dd format.
+        if (claims.containsKey(DATE_OF_BIRTH_LOCAL_CLAIM) &&
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:174]: what happens if the following condition is not matched. then, won't it store again?  We have to throw an error in that case. 

claims.get(DATE_OF_BIRTH_LOCAL_CLAIM).matches(DATE_OF_BIRTH_REGEX)

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java [hunk 2]
```diff
@@ -100,5 +100,10 @@
     public static final String ENABLE_LOGIN_IDENTIFIERS = "LoginIdentifiers.Enable";
     public static final String PRIMARY_LOGIN_IDENTIFIER_CLAIM = "LoginIdentifiers.PrimaryLoginIdentifier";
     public static final boolean DEFAULT_ENABLE_LOGIN_IDENTIFIERS = false;
+
+    // Date of Birth related constants.
+    public static final String DATE_OF_BIRTH_LOCAL_CLAIM = "http://wso2.org/claims/dob";
+    public static final String DATE_OF_BIRTH_REGEX = "^\\d{4}-\\d{2}-\\d{2}$";
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:106]: can't we get this regex from claim metadata?

Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/utils/SCIMCommonConstants.java:107]: addressed in 7b3fb87

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 1 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java]: logic changed in  7b3fb87

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java [hunk 3]
```diff
@@ -165,9 +176,40 @@ public boolean doPreSetUserClaimValuesWithID(String userID, Map<String, String>
         Map<String, String> scimToLocalMappings = SCIMCommonUtils.getSCIMtoLocalMappings();
         String modifiedLocalClaimUri = scimToLocalMappings.get(SCIMConstants.CommonSchemaConstants.LAST_MODIFIED_URI);
         claims.put(modifiedLocalClaimUri, lastModifiedDate);
+
+        // Validate dob value against the regex.
+        validateDateOfBirthClaimValue(claims, userStoreManager);
```
Comment (IsuraD) [inline components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:181]: Let's validate this for doPreSetUserClaimValueWithID as well.

**Review by IsuraD — APPROVED** [review]

File: components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java (hunk 3 above)
Comment (AnuradhaSK) [reply components/org.wso2.carbon.identity.scim2.common/src/main/java/org/wso2/carbon/identity/scim2/common/listener/SCIMUserOperationListener.java:217]: fixed in 3d8d84d

**Pull request merged**